### PR TITLE
[RHPAM-3681] - Add metering labels in KogitoBuild

### PR DESCRIPTION
**Pull request by vaibhavjainwiz**
Many thanks for submitting your Pull Request :heart:! 

Jira issue : https://issues.redhat.com/browse/RHPAM-3681

Please make sure your PR meets the following requirements:

- [x] You have read the [contributors' guide](https://github.com/kiegroup/kogito-operator/blob/master/README.md#contributing-to-the-kogito-operator)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains a link to the JIRA issue
- [x] Pull Request contains a description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster
- [x] You've added a [RELEASE_NOTES.md](RELEASE_NOTES.md) entry regarding this change

### CODE REVIEW

File: core/kogitobuild/build_config.go [hunk 1]
```diff
@@ -17,245 +17,41 @@ package kogitobuild
...
-type DecoratorHandler interface {
-	decoratorForRemoteSourceBuilder() decorator
-	decoratorForLocalSourceBuilder() decorator
-	decoratorForSourceBuilder() decorator
-	decoratorForBinaryRuntimeBuilder() decorator
-	decoratorForSourceRuntimeBuilder() decorator
-	decoratorForRuntimeBuilder() decorator
+// BuildConfigHandler ...
+type BuildConfigHandler interface {
+	NewBuildConfig(build api.KogitoBuildInterface, customLabel map[string]string, decorators ...decorator) buildv1.BuildConfig
```
Comment (sutaakar) [inline core/kogitobuild/build_config.go:32]: Would it be possible and/or meaningful to implement the custom label addition as a decorator?

**Review by sutaakar — APPROVED** [review]

File: core/kogitobuild/build_config.go (hunk 1 above)
Comment (vaibhavjainwiz) [reply core/kogitobuild/build_config.go:32]: Thanks @sutaakar for point it out.
Sure, we could do so. I will implement a new CustomLabelDecorator for it.

File: RELEASE_NOTES.md [hunk 2]
```diff
@@ -3,6 +3,7 @@
 - [KOGITO-4455](https://issues.redhat.com/browse/KOGITO-4455) - Make Kogito Images/CLI to build a Kogito Workflow project from a workflow definition file
 - [KOGITO-4722](https://issues.redhat.com/browse/KOGITO-4722) - Support Job Service MongoDB distribution provisioning in Operator
 - [KOGITO-5315](https://issues.redhat.com/browse/KOGITO-5315) - Jobs service use unsupported authentication mechanism (part of Infinispan 12 upgrade - https://issues.redhat.com/browse/KOGITO-4341)
```
Comment (ricardozanini) [inline RELEASE_NOTES.md:5]: These issues here were already solved, and it's on branch 1.8.x, correct? I think you can remove them.

File: core/kogitobuild/build_config.go [hunk 3]
```diff
@@ -17,245 +17,41 @@ package kogitobuild
...
-
-// newBuildConfig creates a new reference for the very basic default OpenShift BuildConfig reference to build Kogito Services.
+// NewBuildConfig creates a new reference for the very basic default OpenShift BuildConfig reference to build Kogito Services.
 // Pass the required decorator(s) to create the BuildConfig for a particular use case
-func newBuildConfig(build api.KogitoBuildInterface, decorators ...decorator) buildv1.BuildConfig {
-	app := build.GetSpec().GetTargetKogitoRuntime()
-	if len(app) == 0 {
-		app = build.GetName()
-	}
+func (b *buildConfigHandler) NewBuildConfig(build api.KogitoBuildInterface, customLabel map[string]string, decorators ...decorator) buildv1.BuildConfig {
```
Comment (ricardozanini) [inline core/kogitobuild/build_config.go:48]: Why do we need to expose this function? I didn't see a new usage outside of this package.

**Review by ricardozanini — COMMENTED** [review]
Looks good! Just two minor comments. :+1:

**Comment by ricardozanini** [comment]
@vaibhavjainwiz this commit https://github.com/kiegroup/kogito-operator/pull/916/commits/b74b34358e1337c98548cbbba613547309a704f4 fixes the OLM problem

**Review by spolti — APPROVED** [review]
Nice work Vaibhav, 
I assume that @Kevin-Mok will need to update his PR to add the labels to the builds, right?

**Comment by Kevin-Mok** [comment]
@spolti I've tried [implementing this PR](https://github.com/kiegroup/rhpam-kogito-operator/pull/48#issuecomment-871657379) to add the labels to the final build pods, but it doesn't propagate. If adding the labels to the `KogitoBuild`/`BuildConfig`/`Build` is not a priority, I would prefer to merge my existing PR's first so that the labels for the priority runtime pods are added first for the 7.11.1 release, and this can be added at a later date.

It was also originally supposed to be a Canadian holiday for me today, and I took tomorrow off before taking on this task. So preferably, I could get my existing PR's merged first, and then I can submit a new PR for this on Monday after this PR is merged. :sweat_smile:

File: core/kogitobuild/build_config.go (hunk 3 above)
Comment (Kevin-Mok) [reply core/kogitobuild/build_config.go:48]:
```suggestion
func (b *buildConfigHandler) NewBuildConfig(build api.KogitoBuildInterface, customLabels map[string]string, decorators ...decorator) buildv1.BuildConfig {
```

**Review by Kevin-Mok — APPROVED** [review]
Minor nitpick, but otherwise LGTM.

File: core/kogitobuild/build_config.go (hunk 1 above)
Comment (vaibhavjainwiz) [reply core/kogitobuild/build_config.go:32]: done

File: RELEASE_NOTES.md (hunk 2 above)
Comment (vaibhavjainwiz) [reply RELEASE_NOTES.md:5]: done

File: core/kogitobuild/build_config.go (hunk 3 above)
Comment (vaibhavjainwiz) [reply core/kogitobuild/build_config.go:48]: done

**Comment by ricardozanini** [comment]
> @vaibhavjainwiz this commit [b74b343](https://github.com/kiegroup/kogito-operator/commit/b74b34358e1337c98548cbbba613547309a704f4) fixes the OLM problem

@vaibhavjainwiz ^

**Review by radtriste — APPROVED** [review]

**Pull request merged**